src/statsmodels/contrasts.jl

```diff
@@ -0,0 +1,233 @@
+generated (during `ModelFrame` construction).
+
+## Constructors
+
+For `C <: AbstractContrast`:
+
+```julia
+C()                             # levels are inferred later 
+C(levels = ::Vector{Any})       # levels checked against data later
+C(base = ::Any)                 # specify base level
+C(levels = ::Vector{Any}, base = ::Any)
+```
+
+If specified, levels will be checked against data when generating a
+`ContrastsMatrix`. Any mismatch will result in an error, because missing data
+levels would lead to empty columns in the model matrix, and missing contrast
+levels would lead to empty or undefined rows.
+
+You can also specify the base level of the contrasts. The
+actual interpretation of this depends on the particular contrast type, but in
+general it can be thought of as a "reference" level.  It defaults to the first
+level.
+
+Both `levels` and `base` will be coerced to the type of the data when
+constructing a `ContrastsMatrix`.
+
+## Concrete types
+
+* `TreatmentContrasts`
+* `SumContrasts`
+* `HelmertContrasts`
+
+To implement your own concrete types, implement a constructor and a method for
+constructing the actual contrasts matrix that maps from levels to `ModelMatrix`
+column values:
+
+```julia
+type MyContrasts <: AbstractContrasts
+    ...
+end
+
+contrasts_matrix(C::MyContrasts, baseind, n) = ...
+```
+
+"""
+abstract AbstractContrasts
+
+## Contrasts + Data = ContrastsMatrix
+type ContrastsMatrix{T}
+    matrix::Matrix{Float64}
+    termnames::Vector{T}
+    levels::Vector{T}
+    contrasts::AbstractContrasts
+end
+
+"""
+    ContrastsMatrix{T}(::AbstractContrasts, levels::Vector{T})
+
+Instantiate contrasts matrix for given data (categorical levels)
+
+If levels are specified in the `AbstractContrasts`, those will be used, and likewise
+for the base level (which defaults to the first level).
+"""
+function ContrastsMatrix{T}(C::AbstractContrasts, lvls::Vector{T})
+
+    ## if levels are defined on C, use those, validating that they line up.
+    ## what does that mean?
+    ##
+    ## C.levels == lvls (best case)
+    ## data levels missing from contrast: would generate empty/undefined rows. 
+    ## better to filter data frame first
+    ## contrast levels missing from data: would have empty columns, generate a
+    ## rank-deficient model matrix.
+    c_lvls = convert(typeof(lvls), get(C.levels, lvls))
+    mismatched_lvls = symdiff(c_lvls, lvls)
+    isempty(mismatched_lvls) || error("Contrasts levels not found in data or vice-versa: ", mismatched_lvls)
+
+    n = length(c_lvls)
+    n > 1 || error("not enough degrees of freedom to define contrasts")
+    
+    ## find index of base level. use C.base, then default (1).
+    baseind = isnull(C.base) ? 1 : findfirst(c_lvls, convert(eltype(lvls), get(C.base)))
+    baseind > 0 || error("Base level $(C.base) not found in levels")
+    
+    not_base = [1:(baseind-1); (baseind+1):n]
+    tnames = c_lvls[not_base]
```

Comment:
It's correct in the sense that each contrast is interpretable as the difference between that level and the average of the preceding levels. But it might be good to allow custom names (and have a default method that behaves the current way).